crates/ra_hir/src/ty.rs

```diff
@@ -1103,6 +1157,32 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
         let typable = typable?;
         let ty = self.db.type_for_def(typable);
         let ty = self.insert_type_vars(ty);
+
+        // try to get generic parameters from the path and add them to the
+        // function type substitutions
+        if let Ty::FnDef { ref def, .. } = ty {
+            let last_seg_bindings = path
+                .segments
+                .last()
+                .and_then(|segment| segment.args_and_bindings.as_ref());
+            if let Some(generic_args) = last_seg_bindings {
```

Comment:
I think all this should already be done by `substs_from_path`, and it should be enough to call that directly?

Reply:
You are absolutely right - that saves us a lot of work!